File: stdlib/js/hull/middleware/tests/test_idempotency.js
// test_idempotency.js — Tests for hull:middleware:idempotency
//
// Requires db, crypto, time, json globals (run via hull test harness).

import { idempotency } from "hull:middleware:idempotency";
import { db } from "hull:db";
import { json } from "hull:json";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

// ── Helpers ─────────────────────────────────────────────────────────────

function mockRes() {
    const r = {
        _status: 200,
        _body: null,
        _headers: {},
        status(code) { this._status = code; return this; },
        json(data) { this._body = json.encode(data); return this; },
        text(data) { this._body = data; return this; },
        header(name, value) {
            if (value === undefined) return this._headers[name] || null;
            this._headers[name] = value;
            return this;
        },
    };
    return r;
}

function mockReq(method, path, body, headers) {
    const hdrs = headers || {};
    return {
        method: method || "POST",
        path: path || "/api/test",
        body: body || '{"data":"test"}',
        headers: hdrs,
        ctx: {},
        header(name) {
            // Case-insensitive header lookup
            const lower = name.toLowerCase();
            const keys = Object.keys(hdrs);
            for (let i = 0; i < keys.length; i++) {
                if (keys[i].toLowerCase() === lower)
                    return hdrs[keys[i]];
            }
            return null;
        },
    };
}

// ── init ─────────────────────────────────────────────────────────────

test("init creates idempotency table", () => {
    idempotency.init({ ttl: 60 });
});

// ── middleware: no key ───────────────────────────────────────────────

test("no idempotency key proceeds normally", () => {
    const mw = idempotency.middleware();
    const req = mockReq("POST", "/api/test", '{"x":1}');
    const res = mockRes();
    const result = mw(req, res);
    assertEq(result, 0, "should continue");
});

// ── middleware: GET skipped ──────────────────────────────────────────

test("GET requests are not intercepted", () => {
    const mw = idempotency.middleware();
    const req = mockReq("GET", "/api/test", null, { "idempotency-key": "key-get" });
    const res = mockRes();
    const result = mw(req, res);
    assertEq(result, 0);
});

// ── middleware: first request with key ──────────────────────────────

test("first request with key inserts inflight record", () => {
    const mw = idempotency.middleware();
    const req = mockReq("POST", "/api/items", '{"name":"A"}', { "idempotency-key": "js-key-first-1" });
    const res = mockRes();
    const result = mw(req, res);
    assertEq(result, 0, "should continue to handler");
    assertEq(req.ctx._idem_key, "js-key-first-1");
    assertEq(req.ctx._idem_principal, "__anon");

    const rows = db.query(
        "SELECT state FROM _hull_idempotency_keys WHERE key = ?",
        ["js-key-first-1"]
    );
    assertEq(rows.length, 1);
    assertEq(rows[0].state, "inflight");
});

// ── respond + cache hit ──────────────────────────────────────────────

test("respond caches response and retry returns cached", () => {
    const mw = idempotency.middleware();

    // First request
    const req1 = mockReq("POST", "/api/items", '{"name":"B"}', { "idempotency-key": "js-key-cache-1" });
    const res1 = mockRes();
    mw(req1, res1);
    idempotency.respond(req1, res1, 201, { id: 42 });

    // Retry with same key and body
    const req2 = mockReq("POST", "/api/items", '{"name":"B"}', { "idempotency-key": "js-key-cache-1" });
    const res2 = mockRes();
    const result = mw(req2, res2);
    assertEq(result, 1, "should short-circuit");
    assertEq(res2._status, 201, "cached status");
    assertEq(res2._headers["X-Idempotency-Replay"], "true");
});

// ── different body = 409 ─────────────────────────────────────────────

test("same key with different body returns 409", () => {
    const mw = idempotency.middleware();

    // First request
    const req1 = mockReq("POST", "/api/items", '{"name":"C"}', { "idempotency-key": "js-key-conflict-1" });
    const res1 = mockRes();
    mw(req1, res1);
    idempotency.respond(req1, res1, 201, { id: 99 });

    // Retry with same key but different body
    const req2 = mockReq("POST", "/api/items", '{"name":"D"}', { "idempotency-key": "js-key-conflict-1" });
    const res2 = mockRes();
    const result = mw(req2, res2);
    assertEq(result, 1, "should short-circuit");
    assertEq(res2._status, 409, "conflict");
});

// ── inflight = 409 ───────────────────────────────────────────────────

test("concurrent request with same key returns 409", () => {
    const mw = idempotency.middleware();

    // First request (in-flight, not completed)
    const req1 = mockReq("POST", "/api/items", '{"name":"E"}', { "idempotency-key": "js-key-inflight-1" });
    const res1 = mockRes();
    mw(req1, res1);

    // Second request with same key
    const req2 = mockReq("POST", "/api/items", '{"name":"E"}', { "idempotency-key": "js-key-inflight-1" });
    const res2 = mockRes();
    const result = mw(req2, res2);
    assertEq(result, 1, "should short-circuit");
    assertEq(res2._status, 409);
});

// ── complete without respond ─────────────────────────────────────────

test("complete marks key done without caching response", () => {
    const mw = idempotency.middleware();

    const req = mockReq("POST", "/api/items", '{"name":"F"}', { "idempotency-key": "js-key-complete-1" });
    const res = mockRes();
    mw(req, res);
    idempotency.complete(req);

    const rows = db.query(
        "SELECT state, status FROM _hull_idempotency_keys WHERE key = ?",
        ["js-key-complete-1"]
    );
    assertEq(rows.length, 1);
    assertEq(rows[0].state, "complete");
});

// ── principal scoping ────────────────────────────────────────────────

test("different principals can use same key", () => {
    const mw = idempotency.middleware({
        getPrincipal: (req) => req.ctx.userId || "__anon",
    });

    // User A
    const req1 = mockReq("POST", "/api/items", '{"name":"G"}', { "idempotency-key": "js-key-scope-1" });
    req1.ctx.userId = "user-A";
    const res1 = mockRes();
    mw(req1, res1);
    idempotency.respond(req1, res1, 201, { id: 1 });

    // User B with same key
    const req2 = mockReq("POST", "/api/items", '{"name":"G"}', { "idempotency-key": "js-key-scope-1" });
    req2.ctx.userId = "user-B";
    const res2 = mockRes();
    const result = mw(req2, res2);
    assertEq(result, 0, "different principal, should continue");
});

// ── cleanup ─────────────────────────────────────────────────────────

test("cleanup returns count", () => {
    const count = idempotency.cleanup();
    assertEq(typeof count, "number");
});

export default { pass, fail };

File: stdlib/lua/hull/middleware/tests/test_idempotency.lua
-- test_idempotency.lua — Tests for hull.middleware.idempotency
--
-- Requires db, crypto, time, json globals (run via hull test harness).

local idempotency = require('hull.middleware.idempotency')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── Helpers ─────────────────────────────────────────────────────────────

-- Mock response object that captures status and body
local function mock_res()
    local r = {
        _status = 200,
        _body = nil,
        _headers = {},
    }
    function r:status(code) self._status = code; return self end
    function r:json(data) self._body = json.encode(data); return self end
    function r:text(data) self._body = data; return self end
    function r:header(name, value) self._headers[name] = value; return self end
    return r
end

local function mock_req(method, path, body, headers)
    return {
        method = method or "POST",
        path = path or "/api/test",
        body = body or '{"data":"test"}',
        headers = headers or {},
        ctx = {},
    }
end

-- ── init ─────────────────────────────────────────────────────────────

test("init creates idempotency table", function()
    idempotency.init({ ttl = 60 })
end)

-- ── middleware: no key ───────────────────────────────────────────────

test("no idempotency key proceeds normally", function()
    local mw = idempotency.middleware()
    local req = mock_req("POST", "/api/test", '{"x":1}')
    local res = mock_res()
    local result = mw(req, res)
    assert_eq(result, 0, "should continue")
end)

-- ── middleware: GET skipped ──────────────────────────────────────────

test("GET requests are not intercepted", function()
    local mw = idempotency.middleware()
    local req = mock_req("GET", "/api/test", nil, { ["idempotency-key"] = "key-get" })
    local res = mock_res()
    local result = mw(req, res)
    assert_eq(result, 0)
end)

-- ── middleware: first request with key ──────────────────────────────

test("first request with key inserts inflight record", function()
    local mw = idempotency.middleware()
    local req = mock_req("POST", "/api/items", '{"name":"A"}', { ["idempotency-key"] = "key-first-1" })
    local res = mock_res()
    local result = mw(req, res)
    assert_eq(result, 0, "should continue to handler")
    assert_eq(req.ctx._idem_key, "key-first-1")
    assert_eq(req.ctx._idem_principal, "__anon")

    -- Verify inflight record exists
    local rows = db.query(
        "SELECT state FROM _hull_idempotency_keys WHERE key = ?",
        { "key-first-1" }
    )
    assert_eq(#rows, 1)
    assert_eq(rows[1].state, "inflight")
end)

-- ── respond + cache hit ──────────────────────────────────────────────

test("respond caches response and retry returns cached", function()
    local mw = idempotency.middleware()

    -- First request
    local req1 = mock_req("POST", "/api/items", '{"name":"B"}', { ["idempotency-key"] = "key-cache-1" })
    local res1 = mock_res()
    mw(req1, res1)
    idempotency.respond(req1, res1, 201, { id = 42 })

    -- Retry with same key and body
    local req2 = mock_req("POST", "/api/items", '{"name":"B"}', { ["idempotency-key"] = "key-cache-1" })
    local res2 = mock_res()
    local result = mw(req2, res2)
    assert_eq(result, 1, "should short-circuit")
    assert_eq(res2._status, 201, "cached status")
    assert_eq(res2._headers["X-Idempotency-Replay"], "true")
end)

-- ── different body = 409 ─────────────────────────────────────────────

test("same key with different body returns 409", function()
    local mw = idempotency.middleware()

    -- First request
    local req1 = mock_req("POST", "/api/items", '{"name":"C"}', { ["idempotency-key"] = "key-conflict-1" })
    local res1 = mock_res()
    mw(req1, res1)
    idempotency.respond(req1, res1, 201, { id = 99 })

    -- Retry with same key but different body
    local req2 = mock_req("POST", "/api/items", '{"name":"D"}', { ["idempotency-key"] = "key-conflict-1" })
    local res2 = mock_res()
    local result = mw(req2, res2)
    assert_eq(result, 1, "should short-circuit")
    assert_eq(res2._status, 409, "conflict")
end)

-- ── inflight = 409 ───────────────────────────────────────────────────

test("concurrent request with same key returns 409", function()
    local mw = idempotency.middleware()

    -- First request (in-flight, not yet completed)
    local req1 = mock_req("POST", "/api/items", '{"name":"E"}', { ["idempotency-key"] = "key-inflight-1" })
    local res1 = mock_res()
    mw(req1, res1)
    -- Don't call respond — still in-flight

    -- Second request with same key
    local req2 = mock_req("POST", "/api/items", '{"name":"E"}', { ["idempotency-key"] = "key-inflight-1" })
    local res2 = mock_res()
    local result = mw(req2, res2)
    assert_eq(result, 1, "should short-circuit")
    assert_eq(res2._status, 409)
end)

-- ── complete without respond ─────────────────────────────────────────

test("complete marks key done without caching response", function()
    local mw = idempotency.middleware()

    local req = mock_req("POST", "/api/items", '{"name":"F"}', { ["idempotency-key"] = "key-complete-1" })
    local res = mock_res()
    mw(req, res)
    idempotency.complete(req)

    -- Verify state is complete but no status cached
    local rows = db.query(
        "SELECT state, status FROM _hull_idempotency_keys WHERE key = ?",
        { "key-complete-1" }
    )
    assert_eq(#rows, 1)
    assert_eq(rows[1].state, "complete")
end)

-- ── principal scoping ────────────────────────────────────────────────

test("different principals can use same key", function()
    local mw = idempotency.middleware({
        get_principal = function(req) return req.ctx.user_id or "__anon" end,
    })

    -- User A
    local req1 = mock_req("POST", "/api/items", '{"name":"G"}', { ["idempotency-key"] = "key-scope-1" })
    req1.ctx.user_id = "user-A"
    local res1 = mock_res()
    mw(req1, res1)
    idempotency.respond(req1, res1, 201, { id = 1 })

    -- User B with same key
    local req2 = mock_req("POST", "/api/items", '{"name":"G"}', { ["idempotency-key"] = "key-scope-1" })
    req2.ctx.user_id = "user-B"
    local res2 = mock_res()
    local result = mw(req2, res2)
    assert_eq(result, 0, "different principal, should continue")
end)

-- ── cleanup ─────────────────────────────────────────────────────────

test("cleanup returns count", function()
    local count = idempotency.cleanup()
    assert(type(count) == "number", "expected number")
end)

return {pass = pass, fail = fail}
